src/linear_algebra/matrix/sesquilinear_form.lean

```diff
@@ -0,0 +1,640 @@
+begin
+  rw [matrix.to_linear_map₂'_aux, mk₂'ₛₗ_apply],
+  have : (∑ i' j', (if i = i' then 1 else 0) * f i' j' * (if j = j' then 1 else 0)) = f i j :=
+  begin
+    simp_rw [mul_assoc, ←finset.mul_sum],
+    simp only [boole_mul, finset.sum_ite_eq, finset.mem_univ, if_true, mul_comm (f _ _)],
+  end,
+  rw ←this,
+  exact finset.sum_congr rfl (λ _ _, finset.sum_congr rfl (λ _ _, by simp)),
+end
+
+end aux_to_linear_map
+
+section aux_to_matrix
+
+section comm_semiring
+
+variables [comm_semiring R] [comm_semiring R₁] [comm_semiring R₂]
+variables [add_comm_monoid M₁] [module R₁ M₁] [add_comm_monoid M₂] [module R₂ M₂]
+
+variables {σ₁ : R₁ →+* R} {σ₂ : R₂ →+* R}
+
+/-- The linear map from sesquilinear forms to `matrix n m R` given an `n`-indexed basis for `M₁`
+and an `m`-indexed basis for `M₂`.
+
+This is an auxiliary definition for the equivalence `matrix.to_linear_mapₛₗ₂'`. -/
+def linear_map.to_matrix₂_aux (b₁ : n → M₁) (b₂ : m → M₂) :
+  (M₁ →ₛₗ[σ₁] M₂ →ₛₗ[σ₂] R) →ₗ[R] matrix n m R :=
+{ to_fun := λ f, of $ λ i j, f (b₁ i) (b₂ j),
+  map_add' := λ f g, rfl,
+  map_smul' := λ f g, rfl }
+
+@[simp] lemma linear_map.to_matrix₂_aux_apply (f : M₁ →ₛₗ[σ₁] M₂ →ₛₗ[σ₂] R)
+  (b₁ : n → M₁) (b₂ : m → M₂) (i : n) (j : m) :
+  linear_map.to_matrix₂_aux b₁ b₂ f i j = f (b₁ i) (b₂ j) := rfl
+
+end comm_semiring
+
+section comm_ring
+
+variables [comm_ring R] [comm_ring R₁] [comm_ring R₂]
+variables [add_comm_monoid M₁] [module R₁ M₁] [add_comm_monoid M₂] [module R₂ M₂]
+variables [fintype n] [fintype m]
+variables [decidable_eq n] [decidable_eq m]
+
+variables {σ₁ : R₁ →+* R} {σ₂ : R₂ →+* R}
+
+lemma linear_map.to_linear_map₂'_aux_to_matrix₂_aux (f : (n → R₁) →ₛₗ[σ₁] (m → R₂) →ₛₗ[σ₂] R) :
+  matrix.to_linear_map₂'_aux σ₁ σ₂ (linear_map.to_matrix₂_aux
+    (λ i, std_basis R₁ (λ _, R₁) i 1) (λ j, std_basis R₂ (λ _, R₂) j 1) f) = f :=
+begin
+  refine ext_basis (pi.basis_fun R₁ n) (pi.basis_fun R₂ m) (λ i j, _),
+  simp_rw [pi.basis_fun_apply, matrix.to_linear_map₂'_aux_std_basis,
+    linear_map.to_matrix₂_aux_apply],
+end
+
+lemma matrix.to_matrix₂_aux_to_linear_map₂'_aux (f : matrix n m R) :
+  linear_map.to_matrix₂_aux (λ i, std_basis R₁ (λ _, R₁) i 1) (λ j, std_basis R₂ (λ _, R₂) j 1)
+    (f.to_linear_map₂'_aux σ₁ σ₂) = f :=
+by { ext i j, simp_rw [linear_map.to_matrix₂_aux_apply, matrix.to_linear_map₂'_aux_std_basis] }
+
+end comm_ring
+
+end aux_to_matrix
+
+section to_matrix'
+
+/-! ### Bilinear forms over `n → R`
+
+This section deals with the conversion between matrices and sesquilinear forms on `n → R`.
+-/
+
+variables [comm_ring R] [comm_ring R₁] [comm_ring R₂]
+variables [fintype n] [fintype m]
+variables [decidable_eq n] [decidable_eq m]
+
+variables {σ₁ : R₁ →+* R} {σ₂ : R₂ →+* R}
+
+/-- The linear equivalence between sesquilinear forms and `n × m` matrices -/
+def linear_map.to_matrixₛₗ₂' : ((n → R₁) →ₛₗ[σ₁] (m → R₂) →ₛₗ[σ₂] R) ≃ₗ[R] matrix n m R :=
+{ inv_fun := matrix.to_linear_map₂'_aux σ₁ σ₂,
```

Comment:
Usually it's a good idea to explicitly specify data that is available through the `..` operator, in order to ensure it unfolds correctly. So:
```suggestion
{ to_fun := linear_map.to_matrix₂_aux (λ i, std_basis R₁ (λ _, R₁) i 1) (λ j, std_basis R₂ (λ _, R₂) j 1),
  inv_fun := matrix.to_linear_map₂'_aux σ₁ σ₂,
```